File: remote/src/main/java/com/kelsos/mbrc/ui/fragments/MainFragment.java
package com.kelsos.mbrc.ui.fragments;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Typeface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.ShareActionProvider;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.annotations.RepeatMode;
import com.kelsos.mbrc.annotations.ShuffleState;
import com.kelsos.mbrc.domain.TrackPosition;
import com.kelsos.mbrc.dto.track.TrackInfo;
import com.kelsos.mbrc.enums.LfmStatus;
import com.kelsos.mbrc.enums.PlayState;
import com.kelsos.mbrc.presenters.interfaces.IMainViewPresenter;
import com.kelsos.mbrc.ui.dialogs.RatingDialogFragment;
import com.kelsos.mbrc.ui.views.MainView;
import com.kelsos.mbrc.utilities.FontUtils;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnLongClick;
import roboguice.fragment.RoboFragment;

@Singleton public class MainFragment extends RoboFragment implements MainView {
  // Inject elements of the view
  @Bind(R.id.main_artist_label) TextView artistLabel;
  @Bind(R.id.main_title_label) TextView titleLabel;
  @Bind(R.id.main_label_album) TextView albumLabel;
  @Bind(R.id.main_track_progress_current) TextView trackProgressCurrent;
  @Bind(R.id.main_track_duration_total) TextView trackDuration;
  @Bind(R.id.main_button_play_pause) ImageButton playPauseButton;
  @Bind(R.id.main_volume_seeker) SeekBar volumeBar;
  @Bind(R.id.main_track_progress_seeker) SeekBar progressBar;
  @Bind(R.id.main_mute_button) ImageButton muteButton;
  @Bind(R.id.main_shuffle_button) ImageButton shuffleButton;
  @Bind(R.id.main_repeat_button) ImageButton repeatButton;
  @Bind(R.id.main_album_cover_image_view) ImageView albumCover;

  @Inject private IMainViewPresenter presenter;

  private ShareActionProvider mShareActionProvider;

  private Menu menu;
  private SeekBar.OnSeekBarChangeListener volumeBarChangeListener =
      new SeekBar.OnSeekBarChangeListener() {

        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
          if (fromUser) {
            presenter.onVolumeChange(progress);
          }
        }

        public void onStopTrackingTouch(SeekBar seekBar) { }

        public void onStartTrackingTouch(SeekBar seekBar) { }
      };

  private SeekBar.OnSeekBarChangeListener progressBarChangeListener =
      new SeekBar.OnSeekBarChangeListener() {
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
          presenter.onPositionChange(progress);
        }

        public void onStartTrackingTouch(SeekBar seekBar) { }

        public void onStopTrackingTouch(SeekBar seekBar) { }
      };

  private OnExpandToolbarListener listener;

  public MainFragment() {
  }

  @OnClick(R.id.main_button_play_pause) public void playButtonPressed(View v) {
    presenter.onPlayPausePressed();
  }

  @OnClick(R.id.main_button_previous) public void onPreviousButtonPressed(View v) {
    presenter.onPreviousPressed();
  }

  @OnClick(R.id.main_button_next) public void onNextButtonPressed() {
    presenter.onNextPressed();
  }

  @OnLongClick(R.id.main_button_play_pause) public boolean onPlayerStopPressed() {
    presenter.onStopPressed();
    return true;
  }

  @OnClick(R.id.main_mute_button) public void onMuteButtonPressed(View v) {
    presenter.onMutePressed();
  }

  @OnClick(R.id.main_shuffle_button) public void onShuffleButtonClicked(View v) {
    presenter.onShufflePressed();
  }

  @OnClick(R.id.main_repeat_button) public void onRepeatButtonPressed(View v) {
    presenter.onRepeatPressed();
  }

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setHasOptionsMenu(true);
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    final View view = inflater.inflate(R.layout.ui_fragment_main, container, false);
    ButterKnife.bind(this, view);
    presenter.bind(this);

    artistLabel.setSelected(true);
    titleLabel.setSelected(true);
    albumLabel.setSelected(true);

    Typeface robotoRegular = FontUtils.getRobotoRegular(getActivity());
    Typeface robotoMedium = FontUtils.getRobotoMedium(getActivity());

    artistLabel.setTypeface(robotoRegular);
    titleLabel.setTypeface(robotoMedium);
    albumLabel.setTypeface(robotoMedium);
    trackProgressCurrent.setTypeface(robotoMedium);
    trackDuration.setTypeface(robotoMedium);

    progressBar.setOnSeekBarChangeListener(progressBarChangeListener);
    volumeBar.setOnSeekBarChangeListener(volumeBarChangeListener);
    return view;
  }

  @Override public void onPause() {
    super.onPause();
    presenter.onPause();
  }

  @Override public void onResume() {
    super.onResume();
    listener.expandToolbar();
    presenter.onResume();
  }

  @Override public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.menu_lastfm_scrobble:
        presenter.onScrobbleToggle();
        return true;
      case R.id.menu_rating_dialog:
        final RatingDialogFragment ratingDialog = new RatingDialogFragment();
        ratingDialog.show(getActivity().getSupportFragmentManager(), "RatingDialog");
        return true;
      case R.id.menu_lastfm_love:
        presenter.onLfmLoveToggle();
        return true;
      default:
        return false;
    }
  }

  @Override public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
    inflater.inflate(R.menu.menu, menu);
    this.menu = menu;
    MenuItem shareItem = menu.findItem(R.id.actionbar_share);
    mShareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(shareItem);
    mShareActionProvider.setShareIntent(getShareIntent());
  }

  private Intent getShareIntent() {
    Intent shareIntent = new Intent(Intent.ACTION_SEND);
    shareIntent.setType("text/plain");
    final String payload =
        String.format("Now Playing: %s - %s", artistLabel.getText(), titleLabel.getText());
    shareIntent.putExtra(Intent.EXTRA_TEXT, payload);
    return shareIntent;
  }

  @Override public void onAttach(Context context) {
    super.onAttach(context);
    try {
      listener = (OnExpandToolbarListener) context;
    } catch (ClassCastException ex) {
      throw new ClassCastException(context.toString() + "must Implement OnExpandToolbarListener");
    }
  }

  @Override public void updateCover(@Nullable Bitmap bitmap) {
    if (bitmap != null) {
      albumCover.setImageBitmap(bitmap);
    } else {
      albumCover.setImageResource(R.drawable.ic_image_no_cover);
    }
  }

  @Override public void updateShuffle(@ShuffleState String state) {
    int color = getResources().getColor(
        !ShuffleState.OFF.equals(state) ? R.color.accent : R.color.button_dark);
    shuffleButton.setColorFilter(color);

    shuffleButton.setImageResource(
        ShuffleState.AUTODJ.equals(state) ? R.drawable.ic_headset_grey600_24dp
            : R.drawable.ic_shuffle_grey600_24dp);
  }

  @Override public void updateRepeat(@RepeatMode String mode) {
    int color = getResources().getColor(mode.equals(RepeatMode.ALL) ? R.color.accent : R.color.button_dark);
    repeatButton.setColorFilter(color);
  }

  @Override public void updateScrobbling(boolean enabled) {
    final MenuItem scrobbleMenuItem = menu.findItem(R.id.menu_lastfm_scrobble);
    if (scrobbleMenuItem == null) {
      return;
    }
    scrobbleMenuItem.setChecked(enabled);
  }

  @Override public void updateLoved(LfmStatus status) {
    final MenuItem favoriteMenuItem = menu.findItem(R.id.menu_lastfm_love);
    if (favoriteMenuItem == null) {
      return;
    }
    switch (status) {
      case LOVED:
        favoriteMenuItem.setIcon(R.drawable.ic_action_favorite);
        break;
      default:
        favoriteMenuItem.setIcon(R.drawable.ic_action_favorite_outline);
        break;
    }
  }

  @Override public void updateVolume(int volume) {
    volumeBar.setProgress(volume);
  }

  @Override public void updatePlaystate(PlayState playstate) {
    switch (playstate) {
      case PLAYING:
        playPauseButton.setImageResource(R.drawable.ic_pause_circle_fill);
        break;
      case PAUSED:
        playPauseButton.setImageResource(R.drawable.ic_play_circle_fill);
        break;
      case STOPPED:
        playPauseButton.setImageResource(R.drawable.ic_play_circle_fill);
        break;
      default:
        playPauseButton.setImageResource(R.drawable.ic_play_circle_fill);
        break;
    }
  }

  @Override public void updateMute(boolean enabled) {
    muteButton.setImageResource(
        enabled ? R.drawable.ic_volume_off_grey600_24dp : R.drawable.ic_volume_up_grey600_24dp);
  }

  @Override public void updatePosition(TrackPosition position) {
    trackProgressCurrent.setText(String.format("%02d:%02d",
        position.getCurrentMinutes(),
        position.getCurrentSeconds()));

    trackDuration.setText(String.format("%02d:%02d",
        position.getTotalMinutes(),
        position.getTotalSeconds()));

    progressBar.setProgress(position.getCurrent());
    progressBar.setMax(position.getTotal());
  }

  @Override public int getCurrentProgress() {
    return progressBar.getProgress();
  }

  @Override public void setStoppedState() {
    progressBar.setProgress(0);
    trackProgressCurrent.setText("00:00");
  }

  @Override public void updateTrackInfo(TrackInfo info) {
    if (info == null) {
      return;
    }

    artistLabel.setText(info.getArtist());
    titleLabel.setText(info.getTitle());
    albumLabel.setText(TextUtils.isEmpty(info.getYear())
        ? info.getAlbum()
        : String.format("%s [%s]", info.getAlbum(), info.getYear()));

    if (mShareActionProvider != null) {
      mShareActionProvider.setShareIntent(getShareIntent());
    }
  }

  public interface OnExpandToolbarListener {
    void expandToolbar();
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/navigation/main/MainFragment.kt
package com.kelsos.mbrc.ui.navigation.main

import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.ShareActionProvider
import androidx.fragment.app.Fragment
import com.google.android.material.snackbar.Snackbar
import com.kelsos.mbrc.R
import com.kelsos.mbrc.changelog.ChangelogDialog
import com.kelsos.mbrc.content.activestatus.PlayerState
import com.kelsos.mbrc.content.activestatus.PlayerState.State
import com.kelsos.mbrc.content.activestatus.PlayerStatusModel
import com.kelsos.mbrc.content.activestatus.Repeat
import com.kelsos.mbrc.content.activestatus.Repeat.Mode
import com.kelsos.mbrc.content.activestatus.TrackPositionData
import com.kelsos.mbrc.content.activestatus.TrackRatingModel
import com.kelsos.mbrc.content.library.tracks.PlayingTrackModel
import com.kelsos.mbrc.databinding.FragmentMainBinding
import com.kelsos.mbrc.events.ShuffleMode
import com.kelsos.mbrc.events.ShuffleMode.Shuffle
import com.kelsos.mbrc.extensions.getDimens
import com.kelsos.mbrc.networking.connections.Connection
import com.kelsos.mbrc.ui.dialogs.RatingDialogFragment
import com.kelsos.mbrc.ui.navigation.main.ProgressSeekerHelper.ProgressUpdate
import com.squareup.picasso.Picasso
import toothpick.Scope
import toothpick.Toothpick
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MainFragment : Fragment(), MainView, ProgressUpdate {

  @Inject
  lateinit var presenter: MainViewPresenter

  @Inject
  lateinit var progressHelper: ProgressSeekerHelper

  private var mShareActionProvider: ShareActionProvider? = null

  private var changeLogDialog: AlertDialog? = null
  private var outOfDateDialog: AlertDialog? = null

  private var menu: Menu? = null
  private var volumeChangeListener: SeekBarThrottler? = null
  private var positionChangeListener: SeekBarThrottler? = null

  private lateinit var scope: Scope
  private var _binding: FragmentMainBinding? = null
  private val binding get() = _binding!!

  override fun onCreate(savedInstanceState: Bundle?) {
    scope = Toothpick.openScopes(requireActivity().application, PRESENTER_SCOPE, this)
    scope.installModules(MainModule())
    super.onCreate(savedInstanceState)
    Toothpick.inject(this, scope)
  }

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View {
    _binding = FragmentMainBinding.inflate(inflater, container, false)
    return binding.root
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    binding.mainMuteButton.setOnClickListener { presenter.mute() }
    binding.mainShuffleButton.setOnClickListener { presenter.shuffle() }
    binding.mainRepeatButton.setOnClickListener { presenter.repeat() }

    binding.mainButtonPlayPause.setOnClickListener { presenter.play() }
    binding.mainButtonPlayPause.setOnLongClickListener { presenter.stop() }
    binding.mainButtonPrevious.setOnClickListener { presenter.previous() }
    binding.mainButtonNext.setOnClickListener { presenter.next() }

    progressHelper.setProgressListener(this)
    volumeChangeListener = SeekBarThrottler { presenter.changeVolume(it) }
    positionChangeListener = SeekBarThrottler { presenter.seek(it) }
    binding.mainVolumeSeeker.setOnSeekBarChangeListener(volumeChangeListener)
    binding.mainTrackProgressSeeker.setOnSeekBarChangeListener(positionChangeListener)
    binding.mainArtistLabel.isSelected = true
    binding.mainTitleLabel.isSelected = true
    binding.mainAlbumLabel.isSelected = true

    presenter.attach(this)
  }

  override fun onDestroyView() {
    super.onDestroyView()
    _binding = null
  }

  override fun showChangeLog() {
    changeLogDialog = ChangelogDialog.show(requireContext(), R.raw.changelog)
  }

  override fun notifyPluginOutOfDate() {
    val snackBar = Snackbar.make(
      requireView(),
      R.string.main__dialog_plugin_outdated_message,
      Snackbar.LENGTH_INDEFINITE
    )
    snackBar.setAction(android.R.string.ok) { snackBar.dismiss() }
    snackBar.show()
  }

  override fun showPluginUpdateRequired(minimumRequired: String) {
    // TODO re-introduce update required screen
  }

  override fun onStart() {
    super.onStart()

    if (!presenter.isAttached) {
      presenter.attach(this)
    }

    progressHelper.setProgressListener(this)
    volumeChangeListener = SeekBarThrottler { presenter.changeVolume(it) }
    positionChangeListener = SeekBarThrottler { presenter.seek(it) }
    binding.mainVolumeSeeker.setOnSeekBarChangeListener(volumeChangeListener)
    binding.mainTrackProgressSeeker.setOnSeekBarChangeListener(positionChangeListener)
    binding.mainArtistLabel.isSelected = true
    binding.mainTitleLabel.isSelected = true
    binding.mainAlbumLabel.isSelected = true
  }

  override fun onResume() {
    super.onResume()
    presenter.requestNowPlayingPosition()
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    return when (item.itemId) {
      R.id.menu_lastfm_scrobble -> {
        presenter.toggleScrobbling()
        true
      }
      R.id.menu_rating_dialog -> {
        RatingDialogFragment.create(requireActivity() as AppCompatActivity).show()
        true
      }
      R.id.menu_lastfm_love -> {
        presenter.lfmLove()
      }
      else -> false
    }
  }

  override fun onStop() {
    super.onStop()
    presenter.detach()
    progressHelper.setProgressListener(null)
    volumeChangeListener?.terminate()
    volumeChangeListener = null
    positionChangeListener?.terminate()
    positionChangeListener = null
    binding.mainVolumeSeeker.setOnSeekBarChangeListener(null)
    binding.mainTrackProgressSeeker.setOnSeekBarChangeListener(null)
  }

  override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
    super.onCreateOptionsMenu(menu, inflater)
    inflater.inflate(R.menu.menu, menu)
    this.menu = menu
  }

  private val shareIntent: Intent
    get() {
      val shareIntent = Intent(Intent.ACTION_SEND)
      shareIntent.type = "text/plain"
      val payload = String.format(
        "Now Playing: %s - %s",
        binding.mainAlbumLabel.text,
        binding.mainTitleLabel.text
      )
      shareIntent.putExtra(Intent.EXTRA_TEXT, payload)
      return shareIntent
    }

  override fun updateRating(rating: TrackRatingModel) {
    menu?.run {
      val favoriteMenuItem = findItem(R.id.menu_lastfm_love) ?: return

      when (rating.lfmRating) {
        LfmRating.LOVED -> favoriteMenuItem.setIcon(R.drawable.ic_favorite_black_24dp)
        else -> favoriteMenuItem.setIcon(R.drawable.ic_favorite_border_black_24dp)
      }
    }
  }

  override fun updateStatus(playerStatus: PlayerStatusModel) {
    with(playerStatus) {
      updateShuffleState(shuffle)
      updateRepeat(repeat)
      updateVolume(volume, mute)
      updatePlayState(playState)
      updateScrobbleStatus(scrobbling)
    }
  }

  private fun updateCover(path: String) {
    val file = File(path)

    val mainAlbumCoverImageView = binding.mainAlbumCoverImageView
    if (!file.exists()) {
      mainAlbumCoverImageView.setImageResource(R.drawable.ic_image_no_cover)
      return
    }

    if (mainAlbumCoverImageView.tag == path) {
      return
    }

    mainAlbumCoverImageView.tag = path

    val dimens = requireActivity().getDimens()

    Picasso.get()
      .load(file)
      .noFade()
      .error(R.drawable.ic_image_no_cover)
      .config(Bitmap.Config.RGB_565)
      .resize(dimens, dimens)
      .centerCrop()
      .into(mainAlbumCoverImageView)
  }

  private fun updateShuffleState(@Shuffle shuffleState: String) {
    val shuffle = ShuffleMode.OFF != shuffleState
    val autoDj = ShuffleMode.AUTODJ == shuffleState

    val color = requireContext().getColor(if (shuffle) R.color.accent else R.color.button_dark)
    val resId = if (autoDj) R.drawable.ic_headset_black_24dp else R.drawable.ic_shuffle_black_24dp
    binding.mainShuffleButton.setColorFilter(color)
    binding.mainShuffleButton.setImageResource(resId)
  }

  private fun updateRepeat(@Mode mode: String) {
    @ColorRes var colorId = R.color.accent
    @DrawableRes var resId = R.drawable.ic_repeat_black_24dp

    //noinspection StatementWithEmptyBody
    when {
      Repeat.ALL.equals(mode, ignoreCase = true) -> Unit
      Repeat.ONE.equals(mode, ignoreCase = true) -> resId = R.drawable.ic_repeat_one_black_24dp
      else -> colorId = R.color.button_dark
    }

    val repeatButton = binding.mainRepeatButton
    repeatButton.setImageResource(resId)
    repeatButton.setColorFilter(requireContext().getColor(colorId))
  }

  private fun updateVolume(volume: Int, mute: Boolean) {
    binding.mainVolumeSeeker.progress = volume
    val resId = if (mute) {
      R.drawable.ic_volume_off_black_24dp
    } else {
      R.drawable.ic_volume_up_black_24dp
    }
    with(binding.mainMuteButton) {
      setColorFilter(requireContext().getColor(R.color.button_dark))
      setImageResource(resId)
    }
  }

  private fun updatePlayState(@State state: String) {
    val accentColor = requireContext().getColor(R.color.accent)
    val tag = tag(state)

    if (binding.mainButtonPlayPause.tag == tag) {
      return
    }
    val progressBar = binding.mainTrackProgressSeeker
    @DrawableRes val resId: Int = when (state) {
      PlayerState.PLAYING -> {
        /* Start the animation if the track is playing*/
        presenter.requestNowPlayingPosition()
        trackProgressAnimation(progressBar.progress.toLong(), progressBar.max.toLong())
        R.drawable.ic_pause_circle_filled_black_24dp
      }
      PlayerState.PAUSED -> {
        /* Stop the animation if the track is paused*/
        progressHelper.stop()
        R.drawable.ic_play_circle_filled_black_24dp
      }
      PlayerState.STOPPED -> {
        /* Stop the animation if the track is paused*/
        progressHelper.stop()
        activateStoppedState()
        R.drawable.ic_play_circle_filled_black_24dp
      }
      else -> {
        R.drawable.ic_play_circle_filled_black_24dp
      }
    }

    binding.mainButtonPlayPause.setColorFilter(accentColor)
    binding.mainButtonPlayPause.setImageResource(resId)
    binding.mainButtonPlayPause.tag = tag
  }

  /**
   * Starts the progress animation when called. If It was previously running then it restarts it.
   */
  private fun trackProgressAnimation(current: Long, total: Long) {
    progressHelper.stop()

    val tag = binding.mainButtonPlayPause.tag
    if (STOPPED == tag || PAUSED == tag) {
      return
    }

    progressHelper.update(current, total)
  }

  private fun activateStoppedState() {
    binding.mainTrackProgressSeeker.progress = 0
    binding.mainTrackProgressCurrent.text = getString(R.string.playback_progress, 0, 0)
  }

  override fun updateTrackInfo(info: PlayingTrackModel) {
    binding.mainArtistLabel.text = info.artist
    binding.mainTitleLabel.text = info.title
    binding.mainAlbumLabel.text = if (TextUtils.isEmpty(info.year)) info.album
    else String.format("%s [%s]", info.album, info.year)

    if (mShareActionProvider != null) {
      mShareActionProvider!!.setShareIntent(shareIntent)
    }

    updateCover(info.coverUrl)
  }

  override fun updateConnection(status: Int) {
    if (status == Connection.OFF) {
      progressHelper.stop()
      activateStoppedState()
    }
  }

  /**
   * Responsible for updating the displays and seekbar responsible for the display of the track
   * duration and the
   * current progress of playback
   */

  override fun updateProgress(duration: TrackPositionData) {
    updateProgress(duration.current, duration.total)
  }

  private fun updateProgress(
    current: Long,
    total: Long,
  ) {
    var currentSeconds = current / 1000
    var totalSeconds = total / 1000

    val currentMinutes = currentSeconds / 60
    val totalMinutes = totalSeconds / 60

    currentSeconds %= 60
    totalSeconds %= 60
    val finalTotalSeconds = totalSeconds
    val finalCurrentSeconds = currentSeconds

    binding.mainTrackDurationTotal.text = getString(
      R.string.playback_progress,
      totalMinutes,
      finalTotalSeconds
    )
    binding.mainTrackProgressCurrent.text = getString(
      R.string.playback_progress,
      currentMinutes,
      finalCurrentSeconds
    )

    binding.mainTrackProgressSeeker.max = total.toInt()
    binding.mainTrackProgressSeeker.progress = current.toInt()

    trackProgressAnimation(current, total)
  }

  private fun updateScrobbleStatus(active: Boolean) {
    if (menu == null) {
      return
    }
    val scrobbleMenuItem = menu!!.findItem(R.id.menu_lastfm_scrobble) ?: return

    scrobbleMenuItem.isChecked = active
  }

  override fun progress(position: Long, duration: Long) {
    val progressBar = binding.mainTrackProgressSeeker
    val currentProgress = progressBar.progress / 1000
    val currentMinutes = currentProgress / 60
    val currentSeconds = currentProgress % 60

    progressBar.progress = progressBar.progress + 1000
    binding.mainTrackProgressCurrent.text = getString(
      R.string.playback_progress,
      currentMinutes,
      currentSeconds
    )
  }

  override fun onDestroy() {
    Toothpick.closeScope(this)
    outOfDateDialog?.dismiss()
    changeLogDialog?.dismiss()
    super.onDestroy()
  }

  @javax.inject.Scope
  @Target(AnnotationTarget.TYPE)
  @Retention(AnnotationRetention.RUNTIME)
  annotation class Presenter

  companion object {
    private const val PAUSED = "Paused"
    private const val STOPPED = "Stopped"
    private const val PLAYING = "Playing"
    private val PRESENTER_SCOPE: Class<*> = Presenter::class.java

    fun tag(@State state: String): String = when (state) {
      PlayerState.PLAYING -> PLAYING
      PlayerState.PAUSED -> PAUSED
      else -> STOPPED
    }
  }
}
